README.md
WhereTZ: fast and precise timezone lookup
=========================================

**WhereTZ** is a small gem for lookup of timezone by georgraphic
coordinates.

Features:

* quite precise: uses [timezone-boundary-builder](https://github.com/evansiroky/timezone-boundary-builder), which takes source data from OpenStreetMap;
* (relatively) fast: 0.1-0.2 sec for lookup in worst cases and almost immediate lookup for best cases;
* no calls to external services, works without Internet connection;
* no keeping some 50 Mb datafiles in memory or reading them from disk for each call;
* can return timezone name string or `TZInfo::Timezone`.

## Usage

Install it with your usual routines (Gemfile or `gem install`) as `wheretz` gem. Then:

```ruby
require 'wheretz'

WhereTZ.lookup(50.004444, 36.231389) # (lat, lng) order
# => 'Europe/Kiev'

WhereTZ.get(50.004444, 36.231389)
# => #<TZInfo::DataTimezone: Europe/Kiev>

# you should have tzinfo gem installed, wheretz doesn't list it as dependency
```

From command-line, after gem installed:

```bash
wheretz 50.004444,36.231389
# => Europe/Kiev
```

## How it works

1. Latest version of [timezone-boundary-builder](https://github.com/evansiroky/timezone-boundary-builder) dataset is converted into ~400 `data/*.geojson` files;
2. Each of those files corresponds to one timezone; filename contains timezone name and _bounding box_ (min and max latitude and longitude);
3. On each lookup `WhereTZ` first checks provided coordinates by bounding boxes, and if only one bbox (extracted from filename) corresponds to them, returns timezone name immediately;
4. If there's several intersecting bounding boxes, `WhereTZ` reads only relevant timezone files (which are not very large) and checks which polygon actually contains the point.

## Known problems

* Loading/unloading `.geojson` files can be uneffective when called  multiple times; future releases will provide option for preserve data in memory, or for mass lookup of points;
* You should note that gem has ≈50 MiB of datafiles inside;
* Data conversion performed by pretty ugly script (instead of Rake task as it should be).

## Author

[Victor Shepelev](http://zverok.github.io/)

## License

Data license is [ODbL](https://opendatacommons.org/licenses/odbl/).

Code license is usual MIT.


Changelog.md
# WhereTZ changelog

## 0.0.6 (2020-05-02)

* Update timezones info to 2020d (latest available in [
timezone-boundary-builder](https://github.com/evansiroky/timezone-boundary-builder/releases), though there is [2021a](https://www.iana.org/time-zones) already...);
* Slight code style updates.

## 0.0.5 (2020-04-25)

* Update timezones info to 2020a;
* Change timezone files naming to fix problems with zones named like `Asia/Ust-Nera` ([@jotolo](https://github.com/jotolo), [#9](https://github.com/zverok/wheretz/pull/9));
* Remove GeoRuby dependency: after profiling it turned out to be major slowdown, just copy-pasted sole relevant algorithm to work with bare JSON.

## 0.0.4 (2020-04-25)

(unreleased due to some rubygems problem)

* Update timezones info to 2019b (latest available from timezone-boundary-builder)

## 0.0.3 (2019-06-28)

* Data source updated to timezone-boundary-builder (tz_world_map, used previously, is deprecated)
* In (not so rare, as it turned out) case of ambigous timezone, instead of exception, just a list of timezones is returned

## 0.0.2 (2016-03-04)

Fixed edge case (point slightly outside any of polygon).

## 0.0.1 (2016-01-28)

Initial.


spec/spec_helper.rb
# frozen_string_literal: true

# require 'simplecov'

require 'wheretz'
require 'tzinfo'
require 'saharspec'


spec/wheretz_spec.rb
# frozen_string_literal: true

describe WhereTZ do
  describe '#lookup' do
    subject { described_class.method(:lookup) }

    context 'when unambiguous bounding box: Moscow' do
      its_call(55.75, 37.616667) {
        is_expected.to ret('Europe/Moscow').and dont.send_message(File, :read)
      }
    end

    context 'when ambiguous bounding box: Kharkiv' do
      before {
        expect(File).to receive(:read).twice.and_call_original # rubocop:disable RSpec/ExpectInHook,RSpec/MessageSpies
      }

      its_call(50.004444, 36.231389) { is_expected.to ret 'Europe/Kiev' }
    end

    context 'when edge case' do
      its_call(43.6605555555556, 7.2175) { is_expected.to ret 'Europe/Paris' }
    end

    context 'when no timezone: middle of the ocean' do
      its_call(35.024992, -39.481339) { is_expected.to ret nil }
    end

    # FIXME: Can't find new point with overlapping timezones for testing, America/Swift_Current was fixed in 2020a
    xcontext 'when ambiguous timezones' do
      its_call(50.28337, -107.80135) {
        is_expected.to ret ['America/Swift_Current', 'America/Regina']
      }
    end

    context 'when timezone name contains hypen(-) char' do
      its_call(64.56027, 143.22666) { is_expected.to ret 'Asia/Ust-Nera' }
    end
  end

  describe '#get' do
    subject { described_class.method(:get) }

    its_call(55.75, 37.616667) { is_expected.to ret TZInfo::Timezone.get('Europe/Moscow') }
    its_call(35.024992, -39.481339) { is_expected.to ret nil }

    # FIXME: Can't find new point with overlapping timezones for testing, America/Swift_Current was fixed in 2020a
    xcontext 'when ambiguous timezones' do
      its_call(50.28337, -107.80135) {
        is_expected.to ret [TZInfo::Timezone.get('America/Swift_Current'),
                            TZInfo::Timezone.get('America/Regina')]
      }
    end
  end
end


script/simplify.rb
require 'bundler/setup'
require 'progress_bar/core_ext/enumerable_with_progress'
require 'fileutils'
require 'rgeo'
require 'rgeo/geo_json'

TOLERANCE = 0.1 # Don't know, faithfully... Just guessed it or something like this.

all_features = []

Dir['script/data/*.geojson'].each_with_progress do |file|
  collection = RGeo::GeoJSON.decode(File.read(file), json_parser: :json)
  sgeom = collection.first.geometry.simplify_preserve_topology(TOLERANCE)
  sfeature = RGeo::GeoJSON::Feature.new(sgeom, nil, collection.first.properties)
  all_features << sfeature
  scoll = RGeo::GeoJSON::FeatureCollection.new([sfeature])
  File.write(file.sub('script/data/', 'data/'), RGeo::GeoJSON.encode(scoll).to_json)
end

all_coll = RGeo::GeoJSON::FeatureCollection.new(all_features)
File.write('demo/world.geojson', RGeo::GeoJSON.encode(all_coll).to_json)


script/show_boxes.rb
require 'bundler/setup'
require 'rmagick'

class GeoDraw
  def initialize(draw, max_x, max_y)
    @draw, @max_x, @max_y = draw, max_x, max_y
  end

  def rectangle(x1, y1, x2, y2)
    @draw.rectangle(x_geo2img(x1), y_geo2img(y1), x_geo2img(x2), y_geo2img(y2))
  end

  def polygon(*points)
    @draw.polygon(
      *points.each_slice(2).map{|x, y|
        [x_geo2img(x), y_geo2img(y)]
      }.flatten
    )
  end

  include Math
  
  def x_geo2img(lng)
    rescale(lng, -180..180, 0..@max_x) #.tap{|res| p "x: #{lng} => #{res}"}
  end

  def y_geo2img(lat)
    π = PI
    φ = -lat * π / 180
    
    rescale(
      log(tan(π / 4 + φ / 2)),
      -π..π, 0..@max_y
    ) #.tap{|res| p "y: #{lat} => #{res}"}
  end

  def rescale(num, from, to)
    fromdistance = from.end - from.begin
    todistance = to.end - to.begin
    (num - from.begin).to_f / fromdistance * todistance + to.begin
  end
end

class Magick::Draw
  def geo
    @geo or fail("Call Draw#setup_geo beforehands")
  end

  def setup_geo(w, h)
    @geo = GeoDraw.new(self, w, h)
  end
end

include Magick

img = Image.new(1000, 1000){|i| i.background_color = 'white'}
draw = Draw.new.stroke('red').stroke_width(1).fill('transparent')
draw.setup_geo(1000, 1000)

#Dir['data/*.geojson'].each do |f|
Dir['data/Europe*.geojson'].each do |f|
  _, *coords = f.sub('.geojson', '').split('__')
  minx, maxx, miny, maxy = coords.map(&:to_f)
  draw.geo.rectangle(minx, miny, maxx, maxy)
end

draw.draw(img)

img.write('tmp/tzones.png')


script/shp2geojson.rb
require 'bundler/setup'
require 'progress_bar/core_ext/enumerable_with_progress'
require 'geo_ruby'
require 'geo_ruby/shp'

# shapefile = GeoRuby::Shp4r::ShpFile.open('script/data/world/tz_world_mp.shp')
shapefile = GeoRuby::Shp4r::ShpFile.open('script/data/world/combined-shapefile-with-oceans.shp')


shapefile.with_progress.map{|shp|
  bmin, bmax = shp.geometry.bounding_box
  name = shp.data.attributes['TZID']
  next if name == 'uninhabited'

  fname = "data/%s__%.4f__%.4f__%.4f__%.4f.geojson" % [name.gsub('/', '-'), bmin.x, bmax.x, bmin.y, bmax.y]
  File.write(fname,
    {
      "type" => "FeatureCollection",
      "features" => [
        {
          "type" => "Feature",
          "properties" => {'timezone' => shp.data.attributes['TZID']},
          "geometry" => shp.geometry.as_json
        }
      ]
    }.to_json
  )
}


script/split_geojson.rb
require 'bundler/setup'
require 'progress_bar/core_ext/enumerable_with_progress'
require 'geo_ruby'
require 'geo_ruby/geojson'

# shapefile = GeoRuby::Shp4r::ShpFile.open('script/data/world/tz_world_mp.shp')
# shapefile = GeoRuby::Shp4r::ShpFile.open('script/data/world/combined-shapefile-with-oceans.shp')
parser = GeoRuby::GeoJSONParser.new

# TODO: take automatically from https://github.com/evansiroky/timezone-boundary-builder/releases
parser.parse(File.read('script/data/world/combined-with-oceans.json'))

parser.geometry.features.with_progress.each { |feature|
  bmin, bmax = feature.geometry.bounding_box
  name = feature.properties['tzid']
  next if name.start_with?('Etc/') # uninhabited zones, just clutter the data (hopefully)

  fname = "data/%s__%.4f__%.4f__%.4f__%.4f.geojson" % [name.gsub('/', '--'), bmin.x, bmax.x, bmin.y, bmax.y]
  File.write(fname,
    {
      "type" => "FeatureCollection",
      "features" => [
        {
          "type" => "Feature",
          "properties" => {'timezone' => name},
          "geometry" => feature.geometry.as_json
        }
      ]
    }.to_json
  )
}

# shapefile.with_progress.map{|shp|
#   bmin, bmax = shp.geometry.bounding_box
#   name = shp.data.attributes['TZID']
#   next if name == 'uninhabited'

#   fname = "data/%s__%.4f__%.4f__%.4f__%.4f.geojson" % [name.gsub('/', '-'), bmin.x, bmax.x, bmin.y, bmax.y]
#   File.write(fname,
#     {
#       "type" => "FeatureCollection",
#       "features" => [
#         {
#           "type" => "Feature",
#           "properties" => {'timezone' => shp.data.attributes['TZID']},
#           "geometry" => shp.geometry.as_json
#         }
#       ]
#     }.to_json
#   )
# }


lib/wheretz.rb
# frozen_string_literal: true

# WhereTZ is quick and simple time zone lookup by geographic point.
#
# Usage:
#
# ```ruby
# WhereTZ.lookup(50.004444, 36.231389)
# # => 'Europe/Kiev'
#
# WhereTZ.get(50.004444, 36.231389)
# # => #<TZInfo::DataTimezone: Europe/Kiev>
# ```
module WhereTZ
  extend self

  # @private
  FILES =
    Dir[File.expand_path('../data/*.geojson', __dir__)]
    .map { |f|
      name = File.basename(f).sub('.geojson', '')
      zone, *coords = name.split('__')
      zone = zone.gsub('--', '/')
      minx, maxx, miny, maxy = coords.map(&:to_f)
      [f, zone, minx..maxx, miny..maxy]
    }.freeze

  # Time zone name by coordinates.
  #
  # @param lat Latitude (floating point number)
  # @param lng Longitude (floating point number)
  #
  # @return [String, nil, Array<String>] time zone name, or `nil` if no time zone corresponds
  #   to (lat, lng); in rare (yet existing) cases of ambiguous timezones may return an array of names
  def lookup(lat, lng)
    candidates = FILES.select { |_f, _z, xr, yr| xr.cover?(lng) && yr.cover?(lat) }

    case candidates.size
    when 0 then nil
    when 1 then candidates.first[1]
    else
      lookup_geo(lat, lng, candidates)
    end
  end

  # `TZInfo::DataTimezone` object by coordinates.
  #
  # Note that you should add `tzinfo` to your Gemfile to use this method.
  # `wheretz` doesn't depend on `tzinfo` by itself.
  #
  # @param lat Latitude (floating point number)
  # @param lng Longitude (floating point number)
  #
  # @return [TZInfo::DataTimezone, nil, Array<TZInfo::DataTimezone>] timezone object or `nil` if no
  #   timezone corresponds to (lat, lng); in rare (yet existing) cases of ambiguous timezones may
  #   return an array of timezones
  def get(lat, lng)
    begin
      require 'tzinfo'
    rescue LoadError
      raise LoadError, 'Please install tzinfo for using #get'
    end

    name = lookup(lat, lng)
    case name
    when String
      TZInfo::Timezone.get(name)
    when Array
      name.map(&TZInfo::Timezone.method(:get))
    end
  end

  private

  def lookup_geo(lat, lng, candidate_files)
    point = [lng, lat]

    polygons = candidate_files.map { |fname, zone, *| [zone, geom_from_file(fname)] }
    candidates = polygons.select { |_, multipolygon| inside_multipolygon?(multipolygon, point) }

    case candidates.size
    # Since switching to tz-boundary-builder, there should be no "empty" spaces anymore
    when 0 then raise ArgumentError, 'Point outside any known timezone'
    when 1 then candidates.first.first
    else candidates.map(&:first)
    end
  end

  def geom_from_file(fname)
    JSON.parse(File.read(fname)).dig('features', 0, 'geometry')
  end

  def inside_multipolygon?(multipolygon, point)
    polygons(multipolygon).any? { |polygon| contains_point?(polygon, point) }
  end

  # Previously each timezones geojson always contained multypolygon, now it can be just
  # a simple polygon. Make it polymorphic.
  #
  # If it is polygon, ['coordinates'] contains its geometry (array of linear rings, each is array of points)
  # If it is multypolygon, ['coordinates'] is an _array_ of such geometries
  def polygons(geometry)
    case geometry['type']
    when 'Polygon'
      [geometry['coordinates']]
    when 'MultiPolygon'
      geometry['coordinates']
    else
      raise ArgumentError, "Unsupported geometry type: #{geometry['type']}"
    end
  end

  def contains_point?(polygon, (x, y))
    # Taken from GeoRuby's Polygon#contains_point?, which just delegates to LinearRing#contains_point?
    # Polygon's geometry is just an array of linear ring; linear ring is array of points.
    polygon.any? { |points|
      [*points, points.first]
        .each_cons(2)
        .select { |(xa, ya), (xb, yb)|
          (yb > y != ya > y) && (x < (xa - xb) * (y - yb) / (ya - yb) + xb)
        }.size.odd?
    }
  end

  # Last resort: pretty slow check for the cases when the point
  # is slightly outside polygons.
  # See https://github.com/zverok/wheretz/issues/4
  # NB: Not used currently, since switching to timezone-boundary-builder _with oceans_, there are
  # no empty spaces anymore.
  # Left here in case we'll switch to timezones-without-oceans dataset at some point.
  def guess_outside(point, geometries)
    # create pairs [timezone, distance to closest point of its polygon]
    distances = geometries.map { |zone, multipolygon|
      [
        zone,
        polygons(multipolygon).map(&:rings).flatten
                              .map { |p| p.ellipsoidal_distance(point) }.min
      ]
    }

    # FIXME: maybe need some tolerance range for maximal reasonable distance?

    distances.min_by(&:last).first
  end
end
